clickhouse/client.cpp

```diff
@@ -286,29 +289,40 @@ void Client::Impl::Ping() {
 }
 
 void Client::Impl::ResetConnection() {
-    SocketHolder s(SocketConnect(NetworkAddress(options_.host, std::to_string(options_.port))));
+    for (size_t i = 0; i < options_.hosts_ports.size(); ++i) {
+        try {
+            const ClientOptions::HostPort& host_port = options_.hosts_ports[i];
+            SocketHolder s(SocketConnect(NetworkAddress(host_port.host, std::to_string(host_port.port.value_or(options_.port)))));
 
-    if (s.Closed()) {
-        throw std::system_error(errno, std::system_category());
-    }
+            if (s.Closed()) {
+                throw std::system_error(errno, std::system_category());
+            }
 
-    if (options_.tcp_keepalive) {
-        s.SetTcpKeepAlive(options_.tcp_keepalive_idle.count(),
-                          options_.tcp_keepalive_intvl.count(),
-                          options_.tcp_keepalive_cnt);
-    }
-    if (options_.tcp_nodelay) {
-        s.SetTcpNoDelay(options_.tcp_nodelay);
-    }
+            if (options_.tcp_keepalive) {
+                s.SetTcpKeepAlive(options_.tcp_keepalive_idle.count(), options_.tcp_keepalive_intvl.count(), options_.tcp_keepalive_cnt);
+            }
+            if (options_.tcp_nodelay) {
+                s.SetTcpNoDelay(options_.tcp_nodelay);
+            }
+
+            socket_ = std::move(s);
+            socket_input_ = SocketInput(socket_);
+            socket_output_ = SocketOutput(socket_);
+            buffered_input_.Reset();
+            buffered_output_.Reset();
 
-    socket_ = std::move(s);
-    socket_input_ = SocketInput(socket_);
-    socket_output_ = SocketOutput(socket_);
-    buffered_input_.Reset();
-    buffered_output_.Reset();
+            if (!Handshake()) {
+                throw std::runtime_error("fail to connect to " + host_port.host);
+            }
+        } catch (...) {
```

Comment:
Why is it better? We can't predict what type of object will be thrown. Yea, in this function we throw only `runtime_error`, but we don't exactly know what could be thrown from functions, that are used in `ResetConnection`.

Moreover, in `RetryGuard` previous developers [don't catch specific error](https://github.com/ManagedDatabases/clickhouse-cpp/blob/master/clickhouse/client.cpp#L769) --  they also use `...`. It is partly confirms hypothesis, that any object can be thrown from subfunctions of `ResetConnetction`.

Reply:
If we can't predict the type of exception, we don't know the code. But we should do. 
It is better because when you read the code, you understand what you throw.

Reply:
You can catch runtime_error and other exceptions and then use the `...`

Reply:
Done